laiaaa0/assignment3/q1.cpp

```diff
@@ -0,0 +1,51 @@
+
+#include "TestingFunctions.cpp"
+
+int directions_x []= {1,-1,0,0,1,-1,-1,1};
+int directions_y []= {0,0,1,-1,1,1,-1,-1};
+
+bool IsInGrid (int i, int j,const std::vector<std::vector<char> > &grid){
+    if (i>=0 && i<grid.size()){
+      if (j>=0 && j<grid[i].size()) return true;
+    }
+    return false;
+}
+
+void FindAllWordsRecursive (int x, int y, Dictionary * d, std::vector<std::vector<char> > &grid, std::vector<std::vector<bool> > &visited,  std::string current_word,std::vector<std::string> &found_words){
+  current_word = current_word+grid[x][y];
+  visited[x][y] = true;
+  if (d->IsWord(current_word)){
+    found_words.push_back(current_word);
+  }
+  if (d->IsPrefix(current_word)){
```

Comment:
Optional: sometimes adding a lot of nesting (within `if`- or `for`-statements) makes the code harder to read.
In this case you could do the following:

```c++
  // No word can be formed using this prefix, no need to search any further.
  if (!d->IsPrefix(current_word)) return;

   for (int i = 0; i<8; ++i){
     if (IsInGrid(x+directions_x[i],y+directions_y[i],grid) && !visited[x+directions_x[i]][y+directions_y[i]]){
       FindAllWordsRecursive(x+directions_x[i],y+directions_y[i],d,grid,visited,current_word,found_words);
     }
   }
```